Telemetrylib

MCLiii opened the pull request:
Ready to merge.

jrvollmer committed:
Can insert entries into a remote database via REST API calls

There seems to be no noticeable effect on performance from inserting
entries into the database. The dashboard runs just as quickly as before

jrvollmer committed:
Got connection with DataStax Astra working

jrvollmer committed:
Changed back to Oracle connection

jrvollmer committed:
Tested several databases

jrvollmer committed:
Updated submodule

jrvollmer committed:
Added HTTP request to VPS

jrvollmer committed:
VPS is working, and memory issue is fixed

jrvollmer committed:
Cleanup

jrvollmer committed:
Merge branch 'main' of github.com:badgerloop-software/sc1-driver-io into database-testing

jrvollmer committed:
Removed unused resources

MCLiii committed:
TelemetryLibrary

MCLiii committed:
TelemetryLib and program changes

MCLiii committed:
More code to get the library running

MCLiii committed:
changes

MCLiii committed:
Telemetry channel switching is now working

MCLiii committed:
the library appears to be working

MCLiii committed:
change some stuff back

MCLiii committed:
clean up unused variable/func

MCLiii committed:
clean up imports

MCLiii committed:
Delete .vscode directory

MCLiii committed:
Create readme.md

MCLiii committed:
Update readme.md

MCLiii committed:
Update readme.md

MCLiii committed:
Update readme.md

MCLiii committed:
SQL class is working, still need code cleanup

MCLiii committed:
Merge branch 'telemetrylib' of https://github.com/badgerloop-software/sc1-driver-io into telemetrylib

MCLiii committed:
change the site to ping

jrvollmer committed:
Create table when connection is established

MCLiii committed:
implement recievedata, change nullptr return to string to avoid undefined behavior

MCLiii committed:
changed some parameters back

MCLiii committed:
Update readme.md

MCLiii committed:
Update readme.md

jrvollmer commented on DataProcessor/DataGen.h:
```diff
@@ -40,7 +40,7 @@ class DataGen{
     const int SHUTDOWN_RANGE = 600;
     // Upper limit to which the random number is compared for shutdown inputs (randNum < SHUTDOWN_LIMIT -> fault)
-    const int SHUTDOWN_LIMIT = 2;
+    const int SHUTDOWN_LIMIT = 0;
```
Should put this back to 1 or 2 before this is merged into main

MCLiii replied:
yes

jrvollmer commented on DataProcessor/DataUnpacker.cpp:
```diff
@@ -86,23 +86,25 @@ DataUnpacker::DataUnpacker(QObject *parent) : QObject(parent)
 DataUnpacker::~DataUnpacker()
 {
     dataHandlingThread.quit();
+    dataHandlingThread.wait();
```
Why are we waiting after quitting the thread and destroying its parent?

MCLiii replied:
quit tells the thread to stop, but thread could still be running, so use wait to make sure the thread fully stopped before main thread terminates.

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -28,51 +32,44 @@ BackendProcesses::BackendProcesses(QByteArray &bytes, std::vector<std::string> &
+    this->restclient = new QNetworkAccessManager();
+    // TODO Automatically delete server response since it isn't used
+    this->restclient->setAutoDeleteReplies(true);
+    */
```
Can remove non-descriptive comments

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -28,51 +32,44 @@ BackendProcesses::BackendProcesses(QByteArray &bytes, std::vector<std::string> &
+    obj[0]=new SQL(QString::fromStdString(std::to_string(first_msec)));
+    obj[1]=new TCP(QHostAddress::AnyIPv4, 4003);
+    this->tel = new Telemetry(obj);
+    connect(this->tel, &Telemetry::eng_dash_connection, this, &BackendProcesses::comm_status);
```
Could you add one or two comments describing what's being initialized here (52-58) and what the connection is doing?

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -28,51 +32,44 @@ BackendProcesses::BackendProcesses(QByteArray &bytes, std::vector<std::string> &
+
 void BackendProcesses::threadProcedure()
 {
+    if(stop) {
+        return;
+    }
```
What is stop, and why are we returning if it's true?

MCLiii replied:
Stop is the flag that tells the thread to terminate, it's part of thread termination logic to terminate child thread before main thread quits, so I don't get an error every time I closes the program.

jrvollmer replied:
No action needed, but keeping thread unresolved for future clarity

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -106,11 +105,66 @@ void BackendProcesses::threadProcedure()
         //socket->write(QByteArray::fromStdString("Speed: " + std::to_string(speed) + "; Size: " + std::to_string(sizeof(bytes)) + "\n"));
         socket->write(bytes);
     }
+    */
```
Can remove non-descriptive comments

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -106,11 +105,66 @@ void BackendProcesses::threadProcedure()
+    */
+    tel->sendData(bytes);
     mutex.unlock();
     emit dataReady();
```
I know what's going on here, but a brief comment to state what this (especially emitting dataReady()) does would be good

jrvollmer commented on backendprocesses.h:
```diff
@@ -25,31 +24,30 @@ class BackendProcesses : public QObject
 
     QByteArray &bytes;
 
+    std::atomic<bool> stop =false;
```
```suggestion
    std::atomic<bool> stop = false;
```

jrvollmer commented on solar-car-dashboard.pro:
```diff
@@ -5,12 +5,16 @@ CONFIG += c++11
 # You can make your code fail to compile if it uses deprecated APIs.
 # In order to do so, uncomment the following line.
 #DEFINES += QT_DISABLE_DEPRECATED_BEFORE=0x060000    # disables all the APIs deprecated before Qt 6.0.0
+#LIBS += -Ltelemetrylib/cmake-build-debug -llibtelemetrylib
```
If this won't be used to include this addition in the project as a library, the comment can be removed

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,151 @@
+        //      This should help to avoid stalling while waiting for a response from the server
+        //      NOTE - The number of messages sent before a response from the first message is received might still be limited
+        // TODO Send the byte array, along with the corresponding timestamp, to the server
+        // TODO reply = this->restclient->post(request, bytes);
```
Can remove non-descriptive comments

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,151 @@
+
+    std::string receiveData() override{
+        QJsonObject json = QJsonDocument::fromJson(reply->readAll()).object();
+        //QJsonObject json = QJsonDocument::fromJson(QByteArray::fromStdString("JSON Derulo")).object();
```
Can remove non-descriptive comments

MCLiii replied:
Removed

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,151 @@
+    void readReply() override{
+        qDebug()<<"read reply invoked";
+        QJsonObject json = QJsonDocument::fromJson(reply->readAll()).object();
+        //QJsonObject json = QJsonDocument::fromJson(QByteArray::fromStdString("JSON Derulo")).object();
```
Can remove non-descriptive comments

jrvollmer commented on telemetrylib/library.h:
```diff
@@ -0,0 +1,48 @@
+public slots:
+    void comChannelChanged();
+private:
+    //std::stack<QByteArray> dataCache;
```
Can remove non-descriptive comments

MCLiii committed:
implement Dataresend class

MCLiii committed:
Revert "implement Dataresend class"

This reverts commit 6eb25dd3cb38199df76391b015d1e369223862ca.

MCLiii committed:
I'm sorry :<

This reverts commit f9f87ba04b9c9654a237739b0493600a4a8cd19e.

MCLiii committed:
changes to implement resending

jrvollmer commented on telemetrylib/DataResend.cpp:
```diff
@@ -0,0 +1,44 @@
+    qDebug()<<"\n";
+    mutex.unlock();
+    busy = false;
+}
```
Please add some method headers and additional comments in the bodies of the methods.

MCLiii replied:
YES

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,151 @@
+    QString tableToCreate;
+    std::thread *t;
+    std::atomic<bool> finish=false;
+};
```
Please add some method headers and additional comments in the bodies of the methods.

MCLiii replied:
YES

jrvollmer commented on telemetrylib/TCP.cpp:
```diff
@@ -0,0 +1,61 @@
+    bool connected = false;
+};
+//
+
```
Please add some method headers and additional comments in the bodies of the methods.

MCLiii replied:
YES

jrvollmer commented on telemetrylib/library.cpp:
```diff
@@ -0,0 +1,49 @@
+
+void Telemetry::helloworld() {
+    std::cout<<"Helloworld"<<"\n";
+}
```
Two things:
- Please add comments in the bodies of the methods.
- Can remove helloworld()

MCLiii replied:
Si

jrvollmer commented on telemetrylib/library.h:
```diff
@@ -0,0 +1,50 @@
+class Telemetry : public QObject{
+    Q_OBJECT
+public:
+    Telemetry();
+    /**
+     * @param com Data telemetry object ranked by priority
+     * @param size
+     */
+    Telemetry(std::vector<DTI*> comm);
+    /**
+     * to send data, as simple as it gets
+     * @param data
+     */
+    void sendData(QByteArray data);
+    /**
+     * receive data from telemetry
+     * @return data
+     */
+    std::string receiveData();
+    /**
+     * "That's one small step for man, one giant leap for mankind."
+     */
+    void helloworld();
```
Two things:
- I think it would be better to move the method headers to the source file since that's where the meat of the code is.
- Can remove helloworld()

MCLiii replied:
by standard practice headers should be put at where the function is defined, which is in the .h file.
No don't delete my helloworld😭!

jrvollmer replied:
Fair enough, I think these headers can definitely stay. However, I don't think the headers should be neglected in the source files. This is only the declaration of the functions, not the definition, so there is not as much specific detail here. That's in the source files. Things like specific parameter details and return values can be included in the source files
And sorry to rain on the parade, but yes, delete helloworld()

### **EDIT**
Just realized this was for library.h, not DIT.h. My comments above apply to DTI.h, SQL.cpp, and TCP.cpp. Leaving the library function headers in library.h is fine. I think the changes should still be made in DTI.h, SQL.cpp, and TCP.cpp.

jrvollmer commented on telemetrylib/DataResend.h:
```diff
@@ -0,0 +1,35 @@
+
+#include <QObject>
+#include "DTI.h"
+class Dataresend : QObject{
```
```suggestion
class DataResend : QObject{
```

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,151 @@
+            qDebug() << "EMPTY JSON";
+            return "nada";
+        } else {
+            tableName = json.take("response").toString();
+            qDebug() << "HTTP response (table name): " << tableName;
+            return tableName.toStdString();
+            // TODO Automatically delete server responses since they aren't used after reading the table name
+            this->restclient->setAutoDeleteReplies(true);
```
It looks like this isn't what's being used to set the table name (readReply() is used), so setting tableName can be removed or replaced with a more generic code for reading a response.

jrvollmer replied:
See comment in library.cpp, but it looks like this function isn't utilized, so it could be removed entirely

MCLiii replied:
commented out

jrvollmer commented on telemetrylib/library.cpp:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+std::string Telemetry::receiveData() {
+    if (commChannel != -1) {
+        return comm[commChannel]->receiveData();
+    }
+    return "nada";
+}
```
This goes for `receiveData()` in `DTI.h`, `SQL.cpp`, and `TCP.cpp` as well:
- In DTI.h/SQL.cpp/TCP.cpp, it looks like this is being implemented as a method for reading a message that was received from whatever the app is connected to. However, it doesn't look like it's ever actually used since the readReply() slot is used for doing this. I think it would be a good idea to convert receiveData() to a method for explicitly requesting data from whatever the pi is connected to (i.e. give the method an argument for the message to send or the type of data to request from the pi's partner, and the pi will send that request for data). Then, readReply() can still be used to read the response from the pi's partner
- Given the above point: I think this should be renamed to something like `getData()` or `requestData()`

MCLiii replied:
the tcp class has a functional read reply, but sql does not since the slot cleared all the buffers, but since we are not currently reading anything and implementing readReply in sql class would probably need changes in server script, I think we should put aside this task for now.

jrvollmer replied:
I agree, we could leave implementing a getData() function as a separate addition for another task, in which case, **I would either delete the receiveData() function or at the very least comment it out, as it isn't utilized anywhere (except in TCP.cpp, but we aren't reading any messages from the engineering dashboard)**.
Also, SQL has a functional readReply(). It's used to get the table name from the server. However, TCP doesn't have readReply() implemented, though it could be implemented as a slot. I don't see any case where we would explicitly call receiveData() since it's a public function instead of connecting the readReply() slot to read a message as soon as it receives one from the partner host.

jrvollmer replied:
No more action needed, but I'm keeping the thread unresolved for future clarity/reference

jrvollmer commented on telemetrylib/DTI.h:
```diff
@@ -0,0 +1,43 @@
+class DTI : public QObject{
+    Q_OBJECT
+public:
+    /**
+     * Receive bytes via channel to be implemented
+     * @return bytes received
+     */
+    virtual std::string receiveData() = 0;
```
See comment in `library.cpp`, but I think this should be implemented as a method for requesting data rather than reading a response, as the readReply() slot is used to read responses

jrvollmer replied:
See response. Should leave reimplementation as a future task and remove or at least comment out this function.

jrvollmer reviewed (changes requested):
In addition to other suggestions/comments:
- Try running this on a pi and update the project's README with any relevant info (e.g. additional packages we might need to install in the "Compiling and Running the Project on a Rapberry Pi" section). Preferably, try following the README instructions with a different pi that doesn't have any of the dependencies yet. If there are any issues with our current instructions, we can catch them that way.
  - When running it on the pi, as an initial sanity check/stress test, try sending data continuously for 10+ minutes using the cellular connection (can do it with WiFi instead of using the LTE module). Otherwise, check for (... potentially troublesome) memory leaks. I'm just a bit skeptical because allowing HTTP pipelining was required in the LTE testing code to send messages as frequently as we are, but it doesn't seem like it's required with this implementation, which should be pretty much identical to the testing code.
- `library.h/.cpp` should be renamed to something more descriptive (consider `telemetry.h/.cpp` since the class is `Telemetry`).
- Are the changes for breaking stale TCP sockets (Charles' task) in here? **If not, please do NOT add them to this**. I'm just wondering if they're already there so that I know if I should be reviewing them

MCLiii committed:
Update backendprocesses.cpp

remove and added some comments in backend processes

MCLiii committed:
Update backendprocesses.cpp

MCLiii committed:
Update backendprocesses.h

Co-authored-by: jrvollmer <72111610+jrvollmer@users.noreply.github.com>

MCLiii committed:
Update solar-car-dashboard.pro

remove a line of comments

MCLiii committed:
No more comments, the Mingcan style

MCLiii committed:
Update telemetrylib/DataResend.h

Co-authored-by: jrvollmer <72111610+jrvollmer@users.noreply.github.com>

MCLiii committed:
I dont like comments

MCLiii committed:
added comments to the data resend process

MCLiii committed:
change name of the resend class

MCLiii committed:
Comments comments comments

jrvollmer commented on DataProcessor/DataUnpacker.cpp:
```diff
@@ -86,23 +86,25 @@ DataUnpacker::DataUnpacker(QObject *parent) : QObject(parent)
 
 DataUnpacker::~DataUnpacker()
 {
     dataHandlingThread.quit();
+    dataHandlingThread.wait();
```
Could you please add a comment here to explain why we're quitting then waiting? The explanation you gave before would work well: "Quit tells the thread to stop, but thread could still be running, so use wait to make sure the thread fully stopped before main thread terminates."

MCLiii replied:
YES

MCLiii replied:
OKAY

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -28,51 +32,32 @@ BackendProcesses::BackendProcesses(QByteArray &bytes, std::vector<std::string> &
     threadProcedure();
 }
 
+BackendProcesses::~BackendProcesses() {
+    stop=true;
+}
```
Could you add a quick comment like "Set flag to terminate thread procedure before quitting main thread"?

MCLiii replied:
收到🫡

jrvollmer replied:
Still missing comment

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,128 @@
+
+        QUrl myurl;
+        myurl.setScheme("http");
+        myurl.setHost("host"); // TODO
+        myurl.setPort(3000); // TODO
```
You can replace all of these placeholders with the actual IP and port. Also, please remove the TODOs (sorry)

MCLiii replied:
removed

jrvollmer reviewed (changes requested):
Additional changes

MCLiii committed:
Added alot of comments

MCLiii committed:
comments

MCLiii committed:
Update DataGen.h

MCLiii committed:
Comment out receive data

MCLiii committed:
Merge branch 'telemetrylib' of https://github.com/badgerloop-software/sc1-driver-io into telemetrylib

MCLiii committed:
Change Name

MCLiii committed:
Change to accomidate name change

MCLiii committed:
comments

MCLiii committed:
remove todo

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,129 @@
+            this->restclient->setAutoDeleteReplies(true);
+        }
+    }
+    */
```
```suggestion
    /**
     * Read response from the server. Specifically, reads the response to the request to
     * add a new table on the server and sets tableName to the response.
     */
    void readReply() override {
        qDebug()<<"read reply invoked";
        QJsonObject json = QJsonDocument::fromJson(reply->readAll()).object();

        if(json.isEmpty()) {
            qDebug() << "EMPTY JSON";
        } else {
            tableName = json.take("response").toString();
            qDebug() << "HTTP response (table name): " << tableName;

            // TODO Automatically delete server responses since they aren't used after reading the table name
            this->restclient->setAutoDeleteReplies(true);
        }
    }
```
**readReply() has to be included.** This is used to get the table name from the server.
**Also, added a function header.**

jrvollmer commented on telemetrylib/SQL.cpp:
```diff
@@ -0,0 +1,129 @@
+
+        QUrl myurl;
+        myurl.setScheme("http");
+        myurl.setHost("150.136.104.125"); // TODO
+        myurl.setPort(3000); // TODO
+        myurl.setPath("/add-data"); // TODO
+        myurl.setQuery("table-name=" + tableName + "&dataset-time=" + QString::fromStdString(std::to_string(curr_msec)));
+
+        //QNetworkRequest request;
+        request.setUrl(myurl);
+        request.setHeader(QNetworkRequest::ContentTypeHeader, QVariant("arraybuffer")); // TODO Try "blob" for content type as well
```
```suggestion
        myurl.setHost("150.136.104.125");
        myurl.setPort(3000);
        myurl.setPath("/add-data");
        myurl.setQuery("table-name=" + tableName + "&dataset-time=" + QString::fromStdString(std::to_string(curr_msec)));

        //QNetworkRequest request;
        request.setUrl(myurl);
        request.setHeader(QNetworkRequest::ContentTypeHeader, QVariant("arraybuffer"));
```
Again, on me, but the TODOs can be removed

jrvollmer commented on telemetrylib/DataResend.cpp:
```diff
@@ -0,0 +1,44 @@
+}
+
+void DataResend::resend() {
+    busy = true; //if dataresend is currently busy, telemetrylib would not send data directly thru channel, and indead send it thru this class
```
```suggestion
    busy = true; //if DataResend is currently busy, Telemetry would not send data directly thru channel, and instead send it thru this class
```

jrvollmer commented on telemetrylib/DataResend.h:
```diff
@@ -0,0 +1,35 @@
+//
+// Created by Mingcan Li on 2/23/23.
+//
+
+#ifndef TELEMETRYLIB_DATARESEND_H
+#define TELEMETRYLIB_DATARESEND_H
+
+
+#include <QObject>
+#include "DTI.h"
+class DataResend : QObject{
+    Q_OBJECT
+public:
+    void setChannel(DTI *channel);
+    void addToQueue(QByteArray arr);
+    /*
+     * IMPORTANT, tell the class if the telemetry class is connected/disconnected
+     * before giving this class the data
+     */
+    void comStatus(bool state);
+    bool isBusy(){return busy;}
+protected:
+    void resend();
+signals:
+    void send(QByteArray arr);
+private:
+    DTI *channel;
+    QQueue<QByteArray> q;
+    QMutex mutex;
+    std::thread* t;
+    std::atomic<bool> comstate, busy;
+};
+
+
+#endif //TELEMETRYLIB_DATARESEND_H
```
Please add some method headers

MCLiii replied:
好

jrvollmer commented on telemetrylib/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 3.20)
+project(telemetrylib)
+
+set(CMAKE_CXX_STANDARD 11)
```
Did this have to change to run on the pi? I know I ran into some issues that were fixed by changing this to `set(CMAKE_CXX_STANDARD 20)`, but I don't know if that fix was needed for the pi. If it is needed, the change should be made here

MCLiii replied:
Changed, this cmake file compile the telemetrylib itself, but since QT is giving me a hard time importing the library, the code in telemetrylib is currently being compiled as part of the source code of driver io using QMake.

jrvollmer commented on telemetrylib/DataResend.cpp:
```diff
@@ -0,0 +1,44 @@
+    qDebug()<<"\n";
+    mutex.unlock(); //exit the loop data can be added
+    busy = false;
+}
```
It looks like DataResend is properly queueing and dequeuing the byte arrays. However, it's not properly calling sendData() when it dequeues the byte arrays, so the byte arrays aren't being sent to the server

MCLiii replied:
this should be fixed now.

MCLiii committed:
change make version

MCLiii committed:
I can't spell.

Co-authored-by: jrvollmer <72111610+jrvollmer@users.noreply.github.com>

MCLiii committed:
remove todo

Co-authored-by: jrvollmer <72111610+jrvollmer@users.noreply.github.com>

MCLiii committed:
uncomment readReply

Co-authored-by: jrvollmer <72111610+jrvollmer@users.noreply.github.com>

MCLiii committed:
Fixing DataResend

MCLiii committed:
Add comments

MCLiii committed:
added comment to a line of mysterious code

MCLiii committed:
in this folder for a year and we never noticed...

MCLiii committed:
Update C++ version to make atomic work for qmake

MCLiii committed:
Changes to DataResend and DTI to use custom timeStamp

MCLiii committed:
Merge branch 'telemetrylib' of https://github.com/badgerloop-software/sc1-driver-io into telemetrylib

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -28,51 +32,32 @@ BackendProcesses::BackendProcesses(QByteArray &bytes, std::vector<std::string> &
-    _server.listen(QHostAddress::AnyIPv4, 4003);
-    connect(&_server, SIGNAL(newConnection()), this, SLOT(onNewConnection()));
+    obj[0]=new SQL(QString::fromStdString(std::to_string(first_msec))); //This sends data to the cloud server
+    //obj[1]=new TCP(QHostAddress::AnyIPv4, 4003); //this sends data thru TCP sockets
```
Oops. We still need TCP. Also, line 40 will have to be changed to instantiate 2 objs I believe

MCLiii replied:
Accidentally pushed my test code

jrvollmer commented on telemetrylib/Telemetry.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+
+// Send data through the current communication channel
+void Telemetry::sendData(QByteArray bytes) {
+    qDebug() << "send data current comm channel: " << commChannel;
+    // Check if there is a current communication channel and if the resend queue is not busy
+    if (commChannel != -1 && !resendQueue.isBusy()) {
+        // Send data through the current communication channel
+        comm[commChannel]->sendData(bytes);
+    } else {
+        qDebug()<<"adding to queue from telemetry";
+        // Add data to the resend queue if there is no current communication channel or the resend queue is busy
+        resendQueue.addToQueue(bytes);
+    }
+}
```
Let's have this take in a time parameter since we're getting the current time in backendprocesses for setting the `tstamp_xx` fields in the byte array. Then, that time should be passed as an argument to `sendData()` and `addToQueue()` so that we don't have to have special checks in those functions for whether or not a timestamp was passed.

jrvollmer reviewed (changes requested):
Actual refactoring, not just comments

MCLiii committed:
Uncomment TCP instantiation.

MCLiii committed:
Change to make telemetrylib accept timestamp

jrvollmer commented on backendprocesses.cpp:
```diff
@@ -106,11 +93,8 @@ void BackendProcesses::threadProcedure()
-        socket->write(bytes);
-    }
+    // 60 lines of comments were removed here.
+    tel->sendData(bytes); //this passes the data to the telemetrylib to be sent to the chase car
```
Missing curr_msec argument

jrvollmer committed:
Chill out when resending

Added a short sleep during resend to allow the server to receive resent
packets in order. Without the sleep, the server received the first
packet generated after connection was regained in the middle of the
resent packets. The sleep allows packets generated after connection is
regained to be properly queued to be resent after the packets that were
generated with no connection have been sent.

jrvollmer reviewed (approved):
The library works well. There is some work to do for detecting lost TCP connections. However, that's handled by another issue. There is also trouble with resending via TCP. Though that could possibly be handled on the engineering dashboard's end as well. When resending via TCP, the packets are sent too frequently, and the engineering dashboard interprets several packets as the same message, so they cannot be properly unpacked.
